QccPack-0.61-1/src/libQccPackVID/utils/man1/spatialblockencode.1
.TH SPATIALBLOCKENCODE 1 "QCCPACK" ""
.SH NAME
spatialblockencode, spatialblockdecode \- encode/decode 
image sequence using the spatial-block algorithm
.SH SYNOPSIS
.B spatialblockencode 
.RB "[\|" \-fp "\|]"
.RB "[\|" \-hp "\|]"
.RB "[\|" \-qp "\|]"
.RB "[\|" \-ep "\|]"
.RB "[\|" \-f1 
.IR filter1 "\|]"
.RB "[\|" \-f2 
.IR filter2 "\|]"
.RB "[\|" \-f3 
.IR filter3 "\|]"
.RB "[\|" \-w 
.IR wavelet "\|]"
.RB "[\|" \-b
.IR boundary "\|]" 
.RB "[\|" \-nl
.IR num_levels "\|]" 
.RB "[\|" \-mv
.IR mvfile "\|]" 
.RB "[\|" \-rmv "\|]"
.RB "[\|" \-bs
.IR blocksize "\|]" 
.RB "[\|" \-sf
.IR startframe "\|]" 
.RB "[\|" \-ef
.IR endframe "\|]" 
.RB "[\|" \-q "\|]"
.I rate sequence bitstream

.B spatialblockdecode 
.RB "[\|" \-fp "\|]"
.RB "[\|" \-hp "\|]"
.RB "[\|" \-qp "\|]"
.RB "[\|" \-ep "\|]"
.RB "[\|" \-f1 
.IR filter1 "\|]"
.RB "[\|" \-f2 
.IR filter2 "\|]"
.RB "[\|" \-f3 
.IR filter3 "\|]"
.RB "[\|" \-w 
.IR wavelet "\|]" 
.RB "[\|" \-b
.IR boundary "\|]" 
.RB "[\|" \-mv
.IR mvfile "\|]" 
.RB "[\|" \-q "\|]"
.I bitstream sequence
.SH DESCRIPTION
.TP
.B \-fp
Full-pixel accuracy.
.TP
.B \-hp
Half-pixel accuracy.
.TP
.B \-qp
Quarter-pixel accuracy.
.TP
.B \-ep
Eighth-pixel accuracy.
.TP
.BI \-f1 " filter1"
String. Filename of filter to use
for going from full-pixel to half-pixel accuracy.
.TP
.BI \-f2 " filter2"
String. Filename of filter to use
for going from half-pixel to quarter-pixel accuracy.
.TP
.BI \-f3 " filter3"
String. Filename of filter to use
for going from quarter-pixel to eighth-pixel accuracy.
.TP
.BI \-w " wavelet"
String. Name of the wavelet file
.RB ( FBK
or 
.B LFT
for mat); default = "CohenDaubechiesFeauveau.9-7.lft".
.TP
.BI \-b " boundary"
String. Type of boundary extension; default = "symmetric".
.TP
.BI \-nl " num_levels"
Integer. Number of levels of wavelet-transform
decomposition; default value is 3.
.TP
.BI \-mv " mvfile"
String. Name of file to which to read or write motion vectors. Filename
should contain a
.BR printf (3)-style
numerical descriptor which will be filled in with the current frame number.
.TP
.B \-rmw
Read motion vectors from specified file.
.TP
.BI \-bs " blocksize"
Integer. Block size to use in full-search block matching;
default value is 8.
.TP
.BI \-sf " startframe"
Integer. Frame at which to start encoding. Default is first available frame.
.TP
.BI \-ef " endframe"
Integer. Frame at which to end encoding. Default is last available frame.
.TP
.B \-q
Be quiet---do not print any frame statistics to stdout;
default is verbose output.
.SH DESCRIPTION
.B spatialblockencode
encodes an image sequence, 
.IR sequence , 
using the spatial-block algorithm, a generic approach consisting
of spatial-domain motion estimation and compensation followed by
embedded wavelet-based coding of the residual.
.LP
Option
.B \-w
specifies the wavelet, while
option
.B \-b
specifies the boundary extension
for DWT used by the algorithm.
.LP
The desired bitrate of the output
.I bitstream
is specified by
.IR rate ,
which is expressed in bits per pixel of the original image sequence.
.LP
.B spatialblockdecode
performs the corresponding decoding using the spatial-block algorithm, decoding
the
.I bitstream
made by
.B spatialblockencode
and producing
a reconstructed image sequence,
.IR sequence .
For sane results, 
.IR wavelet 
and
.IR boundary
should be the same as they were for
.BR spatialblockencode .
.LP
For both encoding and decoding,
.I sequence
should indicate a collection of grayscale images of the same size stored
as separate, numbered files; the
filename indicated by
.I sequence
must contain one 
.BR printf (3)-style
numerical descriptor which will then be filled in the current
frame number (e.g., football.%03d.pgm will become
football.000.pgm, football.001.pgm, etc.; see
.BR QccPackIMG (3)).
.LP
Both
.B spatialblockencode
and
.B spatialblockdecode
normally print to stdout a small amount of information
after each frame is coded/decoded.
If the 
.B \-q
(quiet) option is given, all printed output is suppressed.
.LP
.B spatialblockencode
and
.B spatialblockdecode
call
.BR QccVIDSpatialBlockEncode (3) 
and 
.BR QccVIDSpatialBlockDecode (3), 
respectively, to perform spatial-block encoding and decoding.
Refer to the man pages of these functions for more information.
.LP
If the
.B \-mv
option is specified to
.BR spatialblockencode ,
motion vectors are written to
.IR mvfile
via
.BR QccVIDMotionVectorsWriteFile (3),
unless
.B \-rmw
is also specified, in which case,
the motion-estimation process of
.B spatialblockencode
is circumvented, and the motion vectors
are read from
.IR mvfile
via
.BR QccVIDMotionVectorsReadFile (3).
For
.BR spatialblockdecode ,
if
.I mvfile
is specified, motion vectors are read from the file
rather than from the encoded bitstream
via
.BR QccVIDMotionVectorsReadFile (3).
.I mvfile
should have a
.BR printf (3)-style
numerical descriptor which will be filled in with
the current frame number before reading or writing
motion vectors.
.SH "SEE ALSO"
.BR QccVIDSpatialBlockEncode (3),
.BR QccVIDSpatialBlockDecode (3),
.BR QccVIDMotionVectorsReadFile (3),
.BR QccVIDMotionVectorsWriteFile (3),
.BR QccPackVID (3),
.BR QccPackSPIHT (3),
.BR QccPackWAV (3),
.BR QccPackIMG (3),
.BR QccPack (3)

S. Cui, Y. Wang, and J. E. Fowler,
"Motion Compensation Via Redundant-Wavelet Multihypothesis,"
.IR "IEEE Transactions on Image Processing" ,
submitted March 2004. Revised February 2005.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackVID/utils/spatialblockencode.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/*
 *  This code was written by Joe Boettcher <jbb15@msstate.edu> based on
 *  the originally developed algorithm and code by Suxia Cui.
 */

#include "spatialblockencode.h"


#define USG_STRING "[-fp %:] [-hp %:] [-qp %:] [-ep %:] [-f1 %: %s:filter1] [-f2 %: %s:filter2] [-f3 %: %s:filter3] [-w %s:wavelet] [-b %s:boundary] [-nl %d:num_levels] [-mv %: %s:mvfile] [-rmv %:] [-bs %d:blocksize] [-sf %: %d:startframe] [-ef %: %d:endframe] [-q %:] %f:rate %s:sequence %s:bitstream"

QccIMGImageSequence ImageSequence;

QccBitBuffer OutputBuffer;

int FullPixelSpecified = 0;
int HalfPixelSpecified = 0;
int QuarterPixelSpecified = 0;
int EighthPixelSpecified = 0;
int MotionAccuracy = QCCVID_ME_FULLPIXEL;

int FilterSpecified1 = 0;
QccString FilterFilename1;
QccFilter Filter1;
int FilterSpecified2 = 0;
QccString FilterFilename2;
QccFilter Filter2;
int FilterSpecified3 = 0;
QccString FilterFilename3;
QccFilter Filter3;

int StartFrameNumSpecified = 0;
int StartFrameNum;
int EndFrameNumSpecified = 0;
int EndFrameNum;

int Quiet = 0;
float Rate = 0.5;
int TargetBitCnt;

int NumRows;
int NumCols;

int NumLevels = 3;

QccWAVWavelet Wavelet;
QccString WaveletFilename = QCCWAVWAVELET_DEFAULT_WAVELET;
QccString Boundary = "symmetric";

int MVFilenameSpecified = 0;
QccString MVFilename;
int ReadMotionVectors = 0;

int Blocksize = 8;


int main(int argc, char *argv[])
{
  FILE *infile;

  QccInit(argc, argv);

  QccWAVWaveletInitialize(&Wavelet);
  QccIMGImageSequenceInitialize(&ImageSequence);
  QccBitBufferInitialize(&OutputBuffer);
  QccFilterInitialize(&Filter1);
  QccFilterInitialize(&Filter2);
  QccFilterInitialize(&Filter3);

  if (QccParseParameters(argc, argv, USG_STRING,
                         &FullPixelSpecified,
                         &HalfPixelSpecified,
                         &QuarterPixelSpecified,
                         &EighthPixelSpecified,
                         &FilterSpecified1,
                         FilterFilename1,
                         &FilterSpecified2,
                         FilterFilename2,
                         &FilterSpecified3,
                         FilterFilename3,
                         WaveletFilename,
                         Boundary, 
                         &NumLevels,
                         &MVFilenameSpecified,
                         &MVFilename,
                         &ReadMotionVectors,
                         &Blocksize,
                         &StartFrameNumSpecified,
			 &StartFrameNum,
                         &EndFrameNumSpecified,
			 &EndFrameNum,
			 &Quiet,
			 &Rate,
			 ImageSequence.filename,
                         OutputBuffer.filename))
    QccErrorExit();

  if (EighthPixelSpecified)
    MotionAccuracy = QCCVID_ME_EIGHTHPIXEL;
  else
    if (QuarterPixelSpecified)
      MotionAccuracy = QCCVID_ME_QUARTERPIXEL;
    else
      if (HalfPixelSpecified)
        MotionAccuracy = QCCVID_ME_HALFPIXEL;
      else
        if (FullPixelSpecified)
          MotionAccuracy = QCCVID_ME_FULLPIXEL;
  
  if (FilterSpecified1)
    {
      if ((infile = QccFileOpen(FilterFilename1, "r")) == NULL)
        {
          QccErrorAddMessage("%s: Error calling QccFileOpen()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccFilterRead(infile, &Filter1))
        {
          QccErrorAddMessage("%s: Error calling QccFilterRead()",
                             argv[0]);
          QccErrorExit();
        }
      QccFileClose(infile);
    }
  if (FilterSpecified2)
    {
      if ((infile = QccFileOpen(FilterFilename2, "r")) == NULL)
        {
          QccErrorAddMessage("%s: Error calling QccFileOpen()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccFilterRead(infile, &Filter2))
        {
          QccErrorAddMessage("%s: Error calling QccFilterRead()",
                             argv[0]);
          QccErrorExit();
        }
      QccFileClose(infile);
    }
  if (FilterSpecified3)
    {
      if ((infile = QccFileOpen(FilterFilename3, "r")) == NULL)
        {
          QccErrorAddMessage("%s: Error calling QccFileOpen()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccFilterRead(infile, &Filter3))
        {
          QccErrorAddMessage("%s: Error calling QccFilterRead()",
                             argv[0]);
          QccErrorExit();
        }
      QccFileClose(infile);
    }

  if (QccWAVWaveletCreate(&Wavelet, WaveletFilename, Boundary))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletCreate()",
                         argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageSequenceFindFrameNums(&ImageSequence))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageSequenceFindFrameNums()",
			 argv[0]);
      QccErrorExit();
    }

  if (StartFrameNumSpecified)
    {
      if (StartFrameNum < ImageSequence.start_frame_num)
        {
          QccErrorAddMessage("%s: Start frame %d is out of the range of available frames which start at %d",
                             argv[0],
                             ImageSequence.start_frame_num);
          QccErrorExit();
        }
      ImageSequence.start_frame_num = StartFrameNum;
    }
  if (EndFrameNumSpecified)
    {
      if (EndFrameNum > ImageSequence.end_frame_num)
        {
          QccErrorAddMessage("%s: End frame %d is out of the range of available frames which end at %d",
                             argv[0],
                             EndFrameNum,
                             ImageSequence.end_frame_num);
          QccErrorExit();
        }
      ImageSequence.end_frame_num = EndFrameNum;
    }

  if (QccIMGImageSequenceStartRead(&ImageSequence))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageSequenceStartRead()",
			 argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageColor(&ImageSequence.current_frame))
    {
      QccErrorAddMessage("%s: Images must be grayscale",
			 argv[0]);
      QccErrorExit();
    }

  if (QccIMGImageGetSize(&ImageSequence.current_frame,
                         &NumRows, &NumCols))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSize()",
			 argv[0]);
      QccErrorExit();
    }

  TargetBitCnt = (int)(Rate * NumRows * NumCols);

  if ((NumRows % Blocksize) ||
      (NumCols % Blocksize))
    {
      QccErrorAddMessage("%s: Image size must be a multiple of %d",
			 argv[0], Blocksize);
      QccErrorExit();
    }

  OutputBuffer.type = QCCBITBUFFER_OUTPUT;
  if (QccBitBufferStart(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferStart()",
                         argv[0]);
      QccErrorExit();
    }
  if (!QccFileSeekable(OutputBuffer.fileptr))
    {
      QccErrorAddMessage("%s: Output bitstream is not a seekable stream",
                         argv[0]);
      QccErrorExit();
    }

  if (ReadMotionVectors && (!MVFilenameSpecified))
    {
      QccErrorAddMessage("%s: Must provide motion-vectors filename for reading",
                         argv[0]);
      QccErrorExit();
    }

  if (QccVIDSpatialBlockEncode(&ImageSequence,
                               ((FilterSpecified1) ?
                                &Filter1 : NULL),
                               ((FilterSpecified2) ?
                                &Filter2 : NULL),
                               ((FilterSpecified3) ?
                                &Filter3 : NULL),
                               MotionAccuracy,
                               &OutputBuffer,
                               Blocksize,
                               NumLevels,
                               TargetBitCnt,
                               &Wavelet,
                               ((MVFilenameSpecified) ?
                                MVFilename : NULL),
                               ReadMotionVectors,
                               Quiet))
    {
      QccErrorAddMessage("%s: Error calling QccVIDSpatialBlockEncode()",
			 argv[0]);
      QccErrorExit();
    }

  if (QccBitBufferEnd(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferEnd()",
                         argv[0]);
      QccErrorExit();
    }

  QccWAVWaveletFree(&Wavelet);
  QccIMGImageSequenceFree(&ImageSequence);
  QccFilterFree(&Filter1);
  QccFilterFree(&Filter2);
  QccFilterFree(&Filter3);

  QccExit;
}
